src/auth/webhook-secret.guard.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import * as crypto from 'crypto';
+import { Request } from 'express';
+import { Env } from 'src/env';
+
+@Injectable()
+export class WebhookSecretGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request: Request = context.switchToHttp().getRequest();
+    const signature = request.headers['x-hub-signature-256'] as string;
+    const secret = Env.SECRET.WEBHOOK;
+    const payload = JSON.stringify(request.body);
+
+    if (!signature) {
+      throw new UnauthorizedException('Missing signature');
+    }
+
+    const hmac = crypto.createHmac('sha256', secret);
+    const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+
+    if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
+      throw new UnauthorizedException('Invalid signature');
+    }
+
+    return true;
+  }
```

Comment:
PullProofAI

The implementation looks good overall. However, there are a few suggestions to enhance the code:
1. Consider adding error handling for potential exceptions during crypto operations.
2. Ensure proper validation of the `Env.SECRET.WEBHOOK` value.
3. Add comments to explain the purpose of each step in the guard.

```diff
+ // Add error handling for crypto operations
+ try {
+     // Crypto operations
+ } catch (error) {
+     throw new UnauthorizedException('Crypto operation failed');
+ }
+
+ // Validate Env.SECRET.WEBHOOK
+ if (!secret) {
+     throw new UnauthorizedException('Webhook secret is missing');
+ }
```

This will improve the robustness and maintainability of the guard implementation.



<!-- This is an auto-generated comment by PullProofAI -->